interpreter/prune.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/common/operators"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+)
+
+type exprPruner struct {
+	expr    *expr.Expr
+	program Program
+	state   EvalState
+}
+
+// Prunes the given expression based on the given EvalState. Modifies the expression in place.
+// A typical use of this interface would be:
+// 1) Evaluate expr with some unknowns,
+// 2) If result is unknown:
+//   a) Maybe clone the expr(typically would be necessary only for the first iteration) and PruneExpr
+//   b) Goto 1
+// Functional call results which are known would be effectively cached across iterations.
+func PruneExpr(expr *expr.Expr, program Program, state EvalState) {
+	pruner := &exprPruner{
+		expr:    expr,
+		program: program,
+		state:   state}
+	pruner.prune(expr)
+}
+
+// Prunes the given expression by evaluating the expression using the constant values in the expression. Modifies the expression in place.
+// A typical use case of this interface would be:
+// 1) Compile the expression (maybe via a service and maybe after checking a compiled expression does not exists in local cache)
+// 2) Prepare the environment and the interpreter
+// 3) ConstantFoldexpr
+// 4) Maybe cache the expression
+//
+// How the environment is prepared in step 2 is flexible. For example, If the caller caches the compiled and constant folded expressions,
+// but is not willing to constant fold( and thus cache results of) some external calls, then he can prepare the overloads accordingly.
+//
+func ConstantFoldExpr(expr *expr.ParsedExpr, interpreter Interpreter, container string) interface{} {
```

Comment:
The concern I have with this function is that it seems as too opinionated about how to fold and makes it appear as though it is doing something fundamentally different from the `PruneExpr` which it isn't. That said, I think there is room for improvement in the making the `WalkExpr()` and `PruneExpr()` do less so that they can be better composed together.

Reply:
Removed this interface.